### ISSUE TITLE
Show overlay of profit/loss in % for each asset in the dashboard

### ISSUE THREAD

**Issue by LefterisJP** [rotki/rotki#701]
## Feature 

**This is a premium only feature**

The dashboard contains an overview of all the assets the user has in his posession and their current USD/EUR/XXX value.

Since Rotki knows about all buys and all sells of a user it can also display the profit/loss in % for the asset at the current point in time. This would be either a green +XX% or a red -XX% next to each asset in the dashboard.

Additionally for users whose jurisdiction considers assets held after XX time as non-taxable this overlay should show how much of the particular asset is able to be sold tax-free.

### Frontend

- Add an overlay over the normal dashboard that would be only available for premium users
- This overlay would show the profit/loss % as described above for each asset.
- If the user has the `taxfree_after_period` setting enabled it would also show how much of the given asset (if any) is able to be sold without any tax considerations.

### Backend

- For this feature to work we would need to have a periodic task tha does the history processing the tax report does, but do it continuously so that the user's data can be queried in a sensible amount of time.
- Add a premium only API endpoint that allows querying the data for the profit/loss overlay for all assets.

**Comment by LefterisJP** [comment]
Very sad to push this to 1.8, but we want v1.7 out rather soon and it can't be with this.

**Comment by adrienlacombe** [comment]
some keywords for the sake of easing search: unrealised unrealized pnl p/l profit profits loss losses

**Comment by adrienlacombe** [comment]
From the mock screenshot above, I wouldn't mind have it a column like "Amount" or "% of net Value", yes be fault it should be from the start of time.

**Comment by Michael-A-Heuer** [comment]
This is a super helpful feature. I had already thought about asking for it.
To give you some feedback as a customer: This feature alone would motivate me for a premium subscription.

In combination with the multiple depot method (#2438, also planned for v1.18.0), the amount of asset that can be sold tax-free should be displayed for each depot separately.

**Comment by adrienlacombe** [comment]
Hi Michael! How are you? I took a premium subscription to help promote this feature :) I definitely agree that it would be super helpful. I don't know what else to do to help, unfortunately I don't know how to dev.

**Comment by LefterisJP** [comment]
> To give you some feedback as a customer: This feature alone would motivate me for a premium subscription.
> Hi Michael! How are you? I took a premium subscription to help promote this feature :)

@Michael-A-Heuer @adrienlacombe 

Feedback heard.
I understand. The problem as usual is each premium customer has different needs and thus asks for different features. Some via emails, some via github others via twitter. We prioritize depending on how much work something requires and how many paying users ask for something. We also have other priorities that change frequently due to responsibilities we have to other businesses.

This issue is non-trivial to implement, hence it gets pushed often.

Doing a +1 on issues you would like helps us. Other than that all that is needed is patience. Software takes time.

> In combination with the multiple depot method (#2438, also planned for v1.18.0),

Mulitple depot is most probably not going in v1.18.0. Only one user has asked for it and would require a huge refactoring. Our milestones are made in a very optimistic way. We are putting things we would like done at some point in milestones and each release we re-evaluate things.

**Comment by anuzis** [comment]
+1 to feature development along these lines, however, rather than a column showing how much of an asset that can be sold tax free ( which only applies to a small handful of countries, and where tax law for those countries is in discussions to change in a few years), it would be more generally useful to all users in all countries to show realized & unrealized gains/losses on each asset. ( example screenshot attached from cointracking )

![RealizedAndUnrealizedGainsAndLosses](https://user-images.githubusercontent.com/1622682/119273876-cdc11f00-bbc1-11eb-999f-d105eb754ebc.png)

I'd like to switch from CoinTracking to Rotki and would gladly continue my premium subscription if this feature can be added. As it stands today, Rotki's free functionality is nice (similar to what zapper.fi provides free), but the premium functionality provides relatively little of value from a tax planning & reporting perspective (candid feedback in the spirit of helping the project improve).

As a new user I may be mistaken, but as far as I can tell the only information Rotki provides (even to my current premium subscription) is a single "Taxable Trade Profit/Loss" value: no difference between short-term Vs long-term gains (broadly relevant to tax law), no breakdown of which asset(s) contributed to short term or long term gains/losses, no breakdown of which unrealized gains/losses are available to help manage end-of-year tax implications, etc.

The foundation of what Rotki has developed is excellent and much of the way there. Looking forward to following closely and continuing the premium subscription if/when more of the above functionality is available.

**Comment by LefterisJP** [comment]
Hey @anuzis thank you for the feedback.

> +1 to feature development along these lines, however, rather than a column showing how much of an asset that can be sold tax free ( which only applies to a small handful of countries, and where tax law for those countries is in discussions to change in a few years), it would be more generally useful to all users in all countries to show realized & unrealized gains/losses on each asset. ( example screenshot attached from cointracking )

Yes the 1 year thing would be customizable. In fact this overlay should eventually be customizable, depending on the user's needs. The idea is you should be able to add/remove any columns you need.

> As a new user I may be mistaken, but as far as I can tell the only information Rotki provides (even to my current premium subscription) is a single "Taxable Trade Profit/Loss" value:

Yes. And it's customizable as much as we can from the accounting settings. The free version has limits in how many events can be processed in the profit/loss report. Once the limit is hit you get a warning.

> Taxable Trade Profit/Loss" value: no difference between short-term Vs long-term gains (broadly relevant to tax law), no breakdown of which asset(s) contributed to short term or long term gains/losses, no breakdown of which unrealized gains/losses are available to help manage end-of-year tax implications, etc.

This short/long term thing is for the US (see: https://github.com/rotki/rotki/issues/2957). The 1 year non-taxable rule is for Germany. Most of our current users are in Germany, thus the 1 year rule calculation is already implemented. But we have gotten requests for adding the short/long term separation too as you can see from the issue.

Rotki's goal is to be as customizable as possible for as many jurisdictions as possible. And do it all in the open, in an opensource project as we believe is a requirement for something that manages your finances. We are a small project with a very small team though. Rotki is doing a lot of stuff. Portfolio tracking in various chains, exchanges, defi protocols, pnl, and tries to help with accounting in various jurisdictions.

That is inherently a really hard thing. Without support from our users we won't be able to grow, and if we can't grow we will not be able to implement complicated features like this "fast enough". Development costs money after all.

> The foundation of what Rotki has developed is excellent and much of the way there. Looking forward to following closely and continuing the premium subscription if/when more of the above functionality is available.

Thank you. We are doing our best. For general feedback join our discord community: https://discord.gg/aGCxHG7
Let's keep this issue clean for discussion on the topic.

**Comment by anuzis** [comment]
@LefterisJP Makes sense, apologies for side tracking this issue with a long winded discussion of other features. Edited my initial comment to shorter suggestion with screenshot. Thanks again & looking forward to following Rotki's development.

**Comment by Zsmith32** [comment]
+1

**Comment by LefterisJP** [comment]
Related issue to add cost basis as an overlay in the dashboard: https://github.com/rotki/rotki/issues/3319

**Comment by aspiers** [comment]
I have just filed #3320 as a follow-on feature to this, to show % of profit/loss not just as a numerical representation of the current value, but via horizon charts which show the historical variance over time.

**Comment by LefterisJP** [comment]
So finally revisiting this.

## Why not implemented yet?

We had a call with the team and decided to push it forward with the target of 1.35.0. The reason this was pushed further and further is how the only way with the current codebase to get this info is by running the super slow and monolithic PnL code. Also adding further automatic processing tasks will hog the 1 process that python has. And we are working on increasing performance and switching to paralelism and multiprocessing but this takes time: https://github.com/rotki/rotki/issues/7906.

## How can we still do it now then?

We will cheat. Instead of happening automatically it will be an ondemand and premium only feature. User will request the processing and calculation per asset and rotki will remember last calculation.

## Specification

### Frontend

- The dashboard should have new columns per asset if premium.
- If not premium it should somehow indicate with a lock that there can be new columns and lead the user to upgrade
- The new columns should be customizable by settings and user can hide any of them. They should be PnL, Average Price. And if setting for 1 year rule sell-free is on, Tax-free amount.
- For each row the user would see either, no result for these columns if task has not ran, or the last value. And user can ask the task to run from the frontend for any asset.
- In the asset page the same information would be requested in a different nicer way and would be queriable there too.

### Backend

- Create a new endpoint to run the asset task. It should accept a list of assets and a boolean to either just query cache or not.
    * If querying cache then it would ask the DB for cached values for the each asset and return if the asset has cached values or not.
    * If querying without cache then **only a single** asset should be accepted. And it should run the task to find the values for the history of this asset as an async task
- The task for finding the value should gather all history events and trades and put them in a list ordered by timestamp, earliest first.
- It should iterate that list and for each event find the pnl, cost basis and calculate the 1 year rule if needed. For trades/swaps find the corresponding in/out event and calculate price from that. For anything else see if it's an in or an out event and use price of asset at the time.
- Shoudl return it via the api but also save it in the DB. Probably in the cache table.
- There should also be some actionable websocket messages if things the use should take action are detected. For example missing acquisition (perhaps we can do a better name?). Essentially if say you are spending 1000 ETH and you only were supposed to have 10 ETH at that point in time.
